druid/src/widget/textbox/editable_text.rs

```diff
@@ -0,0 +1,326 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Traits for text editing and a basic String implementation.
+
+use crate::unicode_segmentation::GraphemeCursor;
+use std::ops::Range;
+
+/// An EditableText trait.
+pub trait EditableText: Sized {
+    type Cursor: EditableTextCursor<Self>;
+
+    /// Create a cursor with a copy of the text and a offset position.
+    fn cursor(&self, position: usize) -> Self::Cursor;
+
+    /// Replace range with new text.
+    fn edit(&mut self, iv: Range<usize>, new: impl Into<String>);
+
+    /// Get slice of text at range.
+    fn slice(&self, iv: Range<usize>) -> Option<&str>;
+
+    /// Get length of text.
+    fn len(&self) -> usize;
+
+    /// Get the next grapheme offset from the given offset, if it exists.
+    fn prev_grapheme_offset(&self, offset: usize) -> Option<usize>;
+
+    /// Get the previous grapheme offset from the given offset, if it exists.
+    fn next_grapheme_offset(&self, offset: usize) -> Option<usize>;
+
+    /// Get the next codepoint offset from the given offset, if it exists.
+    fn prev_codepoint_offset(&self, offset: usize) -> Option<usize>;
+
+    /// Get the previous codepoint offset from the given offset, if it exists.
+    fn next_codepoint_offset(&self, offset: usize) -> Option<usize>;
+}
+
+impl EditableText for String {
+    type Cursor = StringCursor;
+
+    //TODO: this can panic so maybe we should return a Result?
+    fn edit(&mut self, iv: Range<usize>, new: impl Into<String>) {
+        self.replace_range(iv, &new.into());
+    }
+
+    fn slice(&self, iv: Range<usize>) -> Option<&str> {
+        self.get(iv)
+    }
+
+    fn len(&self) -> usize {
+        self.len()
+    }
+
+    fn cursor(&self, position: usize) -> Self::Cursor {
+        // TODO: can we get rid of this .to_string()?
+        // TODO: what happens if this position isn't at a valid offset?
```

Comment:
We have options, I think.

The first option I see is to always advance or retreat to the first valid offset when creating a cursor, to ensure that `cursor.pos()` always returns a valid boundary; in this case we would want to document this explicitly.

The second option is to not validate the initial position, but to explicitly state that a cursor's position is not guaranteed to be valid until one of the cursor movement functions has been called.

So the first is 'easier' but might lead to unexpected behaviour; do we want to move forward or backwards? The second is clearer, but easier to misuse.

Reply:
I ended up returning an Option to give the consumer a chance to react to a cursor being bad.